rfcs/029-work_state_modelling/README.md

```diff
@@ -0,0 +1,177 @@
+
+By separating these two things like this we are able to add new types of work or new pipeline states without worrying about how one will impact the other. The following is a sketch of how this might look:
+
+```scala
+/** WorkState represents the state of the work in the pipeline, and contains
+  * different data depending on what state it is. This allows us to consider the
+  * Work model as a finite state machine with the following stages corresponding
+  * to stages of the pipeline:
+  *
+  *      |
+  *      | (transformer)
+  *      ▼
+  *   Unmerged
+  *      |
+  *      | (matcher / merger)
+  *      ▼
+  *   Merged
+  *      |
+  *      | (relation embedder)
+  *      ▼
+  * Denormalised
+  *      |
+  *      | (id minter)
+  *      ▼
+  *  Identified
+  *
+  * Each WorkState also has an associated IdentifierState which indicates whether
+  * the corresponding WorkData is pre or post the minter.
+  */
+sealed trait WorkState {
+  type Id <: IdentifierState
+  type ImageId <: IdentifierState
+}
+
+object WorkState {
+
+  case class Unmerged() extends WorkState {
+    type Id = IdentifierState.Unminted
+    type ImageId = IdentifierState.Unidentified
+  }
+
+  case class Merged() extends WorkState {
+    type Id = IdentifierState.Unminted
+    type ImageId = IdentifierState.Unidentified
+  }
+
+  case class Denormalised(
+    relations: Relations[Denormalised],
+    sourceIdentifier: SourceIdentifier,
+    identifiedType: String = classOf[Identified].getSimpleName,
+  ) extends WorkState {
+    type Id = IdentifierState.Unminted
+    type ImageId = IdentifierState.Unidentified
+  }
+
+  case class Identified(
+    canonicalId: String,
+    relations: Relations[Identified],
+  ) extends WorkState {
+    type Id = IdentifierState.Minted
+    type ImageId = IdentifierState.Identified
+  }
+}
+
+/** IdentifierState represents the state of individual pieces of work data
+  * (such as genres, concepts etc), with Unminted used to indicate
+  * pre-minter state and Minted used to indicate post-minter state
+  */
+sealed trait IdentifierState
+
+object IdentifierState {
+
+  sealed trait Unminted extends IdentifierState
+  sealed trait Minted extends IdentifierState
+
+  case class Unidentified(
+    sourceIdentifier: SourceIdentifier
+  ) extends Unminted
+
+  case class Identified(
+    sourceIdentifier: SourceIdentifier,
+    canonicalId: String
+  ) extends Minted
+
+  case class Unidentifiable() extends Minted with Unminted
+}
+
+/** Work contains the work itself. It is parameterised by its state, meaning
+  * the same type of Work can be in a number of possible states depending on
+  * where in the pipeline it is. This allows us to easily add new types of work
+  * (such as if Collection is decided to be a separate type to StandardWork),
+  * with the state of the work in the pipeline being an orthogonal concern.
+  */
+sealed trait Work[State <: WorkState] {
```

Comment:
Possibly question, but why is `State` a type parameter on `Work` if it's not used?

Reply:
The subtypes of `Work` all use it, so it enables us to do things like this:

```scala
def getRelations(work: Work[Denormalised]): Relations[Denormalised] =
  work match {
    case StandardWork(_, _, state) => state.relations
    ...
  }
```

Otherwise if we just had a `sealed trait Work {}` every time we wanted to look in the state of one of it's cases we would just have an `Any` which would be pretty useless.

Regardless of access to particular data, it would still be useful for us to have it parameterised by some state due to the FSM part: we can ensure only transitions from one sort of work to another (which wouldn't be possible at the work level if only the sub types were parameterised).

Reply:
For that use case, I’d prefer

```
sealed trait Work[State <: WorkState] {
  …
  val state: State
}

def getRelations(work: Work[Denormalised]): Relations[Denormalised] =
  work.state.relations
```

at which point you delete the method as unnecessary.

Reply:
Yep definitely agree we should do that too